[Frontend] Pythonic tool parser

mdepinet opened the pull request:
This PR adds a tool parser for models that output tools formatted as Python function calls, such as [Llama 3.2](https://github.com/meta-llama/llama-models/blob/main/models/llama3_2/text_prompt_format.md#zero-shot-function-calling) and [ToolACE-8B](https://huggingface.co/Team-ACE/ToolACE-8B).

FIX #9991

mdepinet committed:
pythonic tool parser

Signed-off-by: Mike Depinet <mike@fixie.ai>

DarkLight1337 commented:
cc @K-Mistele

K-Mistele commented:
> cc @K-Mistele

thanks, taking a look!

K-Mistele commented:
Hi @mdepinet, the parser looks good to me! Just a couple notes:

1. Can you include documentation about the newly supported model(s) in the `docs/source/serving/openai_compatible_server.md`? You will see examples for several other model families, so you will just need to adapt it for this parser and maybe indicate which models it is used for.
2. Can you add configurations for the models that you intend for this parser to support to `tests/tool_use/utils.py`, instead of the custom tests that you have built out? I already built out a bunch of parameterized tests for tool-using LLMs and tool parsers, so all you need to do is add configs for the model(s) that this supports. This way, all tool parsers & tool calling models are checked against the same tests. You can find examples of this in the `tests/tool_use/utils.py` file itself, or you can see an example of it [here](https://github.com/vllm-project/vllm/pull/8405/files#diff-09edf1d6b6a0b2399b64b29fe8a446cb5c75fe107c0ad94c0b79585abef5f228) in a PR for another model's tool parser.
Once you've done that, you can run the tests locally using `pytest tests/tool_use` to make sure they pass, and you should be able to use the `-k` flag to select the parameterized versions for your model only if you desire.

mdepinet committed:
Add an entry to openai_compatible_server.md

Signed-off-by: Mike Depinet <mike@fixie.ai>

mdepinet commented:
> Hi @mdepinet, the parser looks good to me! Just a couple notes:
> 
> 1. Can you include documentation about the newly supported model(s) in the `docs/source/serving/openai_compatible_server.md`? You will see examples for several other model families, so you will just need to adapt it for this parser and maybe indicate which models it is used for.

Sure, done.

> 2. Can you add configurations for the models that you intend for this parser to support to `tests/tool_use/utils.py`, instead of the custom tests that you have built out? I already built out a bunch of parameterized tests for tool-using LLMs and tool parsers, so all you need to do is add configs for the model(s) that this supports. This way, all tool parsers & tool calling models are checked against the same tests. You can find examples of this in the `tests/tool_use/utils.py` file itself, or you can see an example of it [here](https://github.com/vllm-project/vllm/pull/8405/files#diff-09edf1d6b6a0b2399b64b29fe8a446cb5c75fe107c0ad94c0b79585abef5f228) in a PR for another model's tool parser.
>    Once you've done that, you can run the tests locally using `pytest tests/tool_use` to make sure they pass, and you should be able to use the `-k` flag to select the parameterized versions for your model only if you desire.

Thanks, I didn't see these.  I'm happy to add a config so these tests exercise the new tool parser also, but I think it's worthwhile to keep the tests I added also because:

1. The new tests are **much** faster to run, which is quite helpful when iterating on the implementation code.  They're also easier to get running since they have fewer dependencies (no separate server).
1. The new tests don't depend on any particular model or chat template.  It makes sense to tie everything together when a strict JSON schema implies that tight coupling, but I think it's reasonable to define (and test) how the parser works in isolation for a parser that may have broader applicability.
1. The new tests cover more cases than what would be reasonable in the existing tests.

Put differently, I view the existing tests more as integration tests and the tests I added more as unit tests.  I think it'd be a mistake to entirely do away with either.  Does that seem reasonable to you?

mdepinet committed:
Integration test attempt (can't run on my poor 4070)

Signed-off-by: Mike Depinet <mike@fixie.ai>

mdepinet committed:
checkpoint: fix most tool_use tests

Signed-off-by: Mike Depinet <mike@fixie.ai>

K-Mistele commented:
> * The new tests are **much** faster to run, which is quite helpful when iterating on the implementation code.  They're also easier to get running since they have fewer dependencies (no separate server).

For sure, having extra ones is great! I just wanted to make sure that we didn't skip adding this into the existing integration tests for tool use. Checking it out now for testing :)

cc @maxdebayser it sounds like this might handle the llama 3.2 1B and 3B tool calling format that we were having issues with , using the `llama3_json` parser. 

Possible closes #9991 ? will test.

K-Mistele commented:
Not an issue with this PR, but `Team-ACE/ToolACE-8B` requires a special chat template that the model's repo doesn't provide (or additional prompting techniques), in order for the model to properly use tools.

Couple options here:
1. We can remove `Team-ACE/ToolACE-8B` from mentions of the `pythonic` tool parser until the chat template is added in `examples` (which i could create a separate PR for) 
2. We can add a chat template and indicate in the docs that people need to use it, akin to how we provided recommended chat templates for some other models.

I'll provide a tool call parser here in a bit if you're willing to add it to `examples/tool_ace_chat_template.jinja` and indicate that the model requires it to use tools properly, absent complicated prompting that must be done by the end-user.

K-Mistele commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -274,6 +274,27 @@ Flags: `--tool-call-parser granite-20b-fc --chat-template examples/tool_chat_tem
+* Llama's smaller models struggle to use tools effectively.
+
+Example supported models:
+* `meta-llama/Llama-3.2-1B-Instruct`
+* `meta-llama/Llama-3.2-3B-Instruct`
```
Can you indicate that these models need `--chat-template examples/tool_chat_template_llama3.2_pythonic.jinja` to behave as expected?

mdepinet replied:
Done.

K-Mistele commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -274,6 +274,27 @@ Flags: `--tool-call-parser granite-20b-fc --chat-template examples/tool_chat_tem
+Example supported models:
+* `meta-llama/Llama-3.2-1B-Instruct`
+* `meta-llama/Llama-3.2-3B-Instruct`
+* `Team-ACE/ToolACE-8B`
+
+Flags: `--tool-call-parser pythonic`
```
I can provide the chat template for ToolACE but this also requires a special chat template to behave properly

mdepinet replied:
Updated. Also went ahead and added that template.

K-Mistele commented on tests/tool_use/utils.py:
```diff
@@ -111,7 +111,18 @@ def ensure_system_prompt(messages: List[Dict[str, Any]],
         False,
-    }
+    },
+    "pythonic": {
```
can you name this one `llama3.2` and create a separate one for Tool ACE once the chat template is ready?

mdepinet replied:
Done.

K-Mistele commented:
using `vllm serve meta-llama/Llama-3.2-3B-Instruct --enable-auto-tool-choice --tool-call-parser pythonic --chat-template examples/tool_chat_template_llama3.2_pythonic.jinja` I had a really hard time getting a valid tool call out of the model even at a `0` temperature. Unclear if it's just because the model is really small or if it needs some better prompting on my end, but here are a few examples of what I got: 

![Screenshot 2024-11-06 at 11 52 27 PM](https://github.com/user-attachments/assets/0221f607-da23-4a43-aad7-120bd6380a5f)
![Screenshot 2024-11-06 at 11 51 55 PM](https://github.com/user-attachments/assets/bdbffda6-5031-499a-b719-c62ca7399c29)

The first one looks like it _could_ be a parser error, but I'm not sure.

K-Mistele commented:
Just tried running tests with pytest - everything looks good, the model just fails the tests with some frequency (`pytest -v -s tests/tool_use -k pythonic` will select only the tests for this tool parser). 

@mdepinet can you see if it's possible to get them passing with some additional prompting? I think `temperature=0` is default for these tests so that should help some.

mdepinet committed:
Get remaining tool_use tests passing

Signed-off-by: Mike Depinet <mike@fixie.ai>

mdepinet committed:
Merge remote-tracking branch 'origin/main' into mike/pythonic-tool-calls

mdepinet committed:
Add ToolACE template and tool_use test entry

Signed-off-by: Mike Depinet <mike@fixie.ai>

mdepinet committed:
update docs

Signed-off-by: Mike Depinet <mike@fixie.ai>

mdepinet commented:
@K-Mistele I think this is ready for you now.  The smaller Llama models aren't especially reliable.  The tests are passing for me, but I'd be inclined to remove the Llama3.2 entry in favor of ToolACE if it's still flaky for you.  (I'm actually most interested in fixie-ai/ultravox-v0_4-ToolACE-8B anyway.)

K-Mistele commented:
Thanks for the heads-up, I will give it another pass!

K-Mistele commented on docs/source/serving/openai_compatible_server.md:
````diff
@@ -315,6 +315,28 @@ AI21's Jamba-1.5 models are supported.
+```
+
+Limitations:
+* The model must not generate both text and tool calls in the same generation. This may not be hard to change for a specific model, but the community currently lacks consensus on which tokens to emit when starting and ending tool calls.  (In particular, the Llama 3.2 models emit no such tokens.)
````
I've seen the llama 3.2 models respond with the `<|python_tag|>` token at the beginning which is not handled correctly, see the screenshots earlier in the thread

mdepinet replied:
Since option 2 below seems to be the consensus, I'm not planning on handling `<|python_tag|>` right now, but I've added a TODO.

K-Mistele commented:
Yeah so overall, ToolACE implementation is definitely good to go. The llama 3.2 models I think need the `<|python_tag|>` issue I called out above handled, but I could be wrong. I see valid tool calls about half the time. The ones that aren't usually look like this:

![Screenshot 2024-11-07 at 9 37 27 PM](https://github.com/user-attachments/assets/5cc0dada-fed4-44be-8a87-f8c95b4fcdf6)
(`<|python_tag|>` but no array wrapping the calls) 

or like this:
![Screenshot 2024-11-07 at 9 38 09 PM](https://github.com/user-attachments/assets/70247030-6c4e-4cf4-ba22-6751d3fcca5f)
(weirdly, the model is telling you what tools to call. I saw this with 3.1 some, which led to me creating my own chat template with a custom system prompt that fixes a lot of these issues; feel free to borrow/adapt the prompt if you want: https://gist.github.com/K-Mistele/820d142b4dab50bd8ef0c7bbcad4515c)

For what it's worth, some parsers like `hermes` to handle cases where the model generates text and a tool call; I think how you would handle this one for streaming is look for `\n\n[` since that's what the model does whenever it wants to talk and call a tool - there's text, then a double newline and the array. Hermes's parser is a good example of streaming both text and tool deltas depending on what's being generated. 

I think there are a couple ways to move forwards here:
- move forwards with just `ToolACE` and the `fixie-ai` model; removing Llama 3.2 1B and 3B
- include the llama 3.2 models in the PR, but note the low quality and omit them from CI since they are likely to fail
- include the llama 3.2 models in the PR, but try to fix behavior with system prompts for when they're run in CI in `tests/tool_use/utils.py` by telling them to only generate tool calls and don't generate text if they're generating a tool call (the default system prompts & behavior from the llama 3.1 and 3.2 models is notoriously bad and this would probably go a long way to fixing them)

What do you think @mdepinet @DarkLight1337 @mgoin ?

DarkLight1337 commented:
>  I think there are a couple ways to move forwards here:
> 
> * move forwards with just `ToolACE` and the `fixie-ai` model; removing Llama 3.2 1B and 3B
> * include the llama 3.2 models in the PR, but note the low quality and omit them from CI since they are likely to fail
> * include the llama 3.2 models in the PR, but try to fix behavior with system prompts for when they're run in CI in `tests/tool_use/utils.py` by telling them to only generate tool calls and don't generate text if they're generating a tool call (the default system prompts & behavior from the llama 3.1 and 3.2 models is notoriously bad and this would probably go a long way to fixing them)

Unless the author of #9991 @SinanAkkoyun is ok with the first option, I'd go with the second option, and perhaps also log a warning in the code for Llama 3.2 models.

SinanAkkoyun commented:
Hi!
@DarkLight1337 https://github.com/vllm-project/vllm/issues/9991#issuecomment-2464223088 Ollama seems to handle llama3.2 function calling good

However, I don't really care if llama3.2 tool calling is supported here as long as I can verify that for example the hermes tool calling format works, but I can't even verify that sadly. Maybe I am missing configs?

mdepinet committed:
Warn about Llama3.2 models, add TODO for future work

Signed-off-by: Mike Depinet <mike@fixie.ai>

DarkLight1337 commented on docs/source/serving/openai_compatible_server.md:
```diff
@@ -329,13 +329,15 @@ Limitations:
 
 Flags: `--tool-call-parser pythonic --chat-template {see_above}`
 
+\* Llama's smaller models frequently fail to emit tool calls in the correct format. Your mileage may vary.
```
Let's use a warning block to be more obvious.

mdepinet replied:
Done.

DarkLight1337 commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -75,6 +75,11 @@ def __init__(self,
             try:
                 self.tool_parser = ToolParserManager.get_tool_parser(
                     tool_parser)
+                if (self.tool_parser.__name__ == "PythonicToolParser" and
```
Perhaps it would be better to match against `tool_parser` variable instead?

mdepinet replied:
:man_facepalming: Yes, of course. Done.

K-Mistele commented:
Related #10164

mdepinet committed:
PR comments

Signed-off-by: Mike Depinet <mike@fixie.ai>

mdepinet committed:
Alter warning block based on other "note" block in this file

Signed-off-by: Mike Depinet <mike@fixie.ai>

DarkLight1337 reviewed (approved):
Thanks for adding this!
